text/0192-remove-constructs-compat.md

```diff
@@ -0,0 +1,667 @@
+   mechanical fixes and easy to do.
+2. Add some automation using eslint to fix these merge conflicts.
+3. Perform the change on the 1.x codebase such that the soon-to-be-moved types
+   are imported separately from other types of `@aws-cdk/core`. This will
+   violate an eslint rule, but we can disable it for the time being.
+
+We can start with #1 for a while and see how bad it is. If it is, we can try to
+introduce some automation or back-porting.
+
+Something that we can definitely do on 1.x is replace all the variants such as
+`core.Construct` and `cdk.Construct` to just `Construct`. This will slightly
+help.
+
+
+## Removal of "synthesize"
+
+Version 4.x of the `constructs` library does not contain a lifecycle hook for
+synthesis.
+
+### Motivation
+
+The reason this is not available at the base class is because the abstraction
+did not "hold water" as the AWS CDK evolved and new CDKs emerged.  In the AWS
+CDK, we eventually ended up with a centralized synthesis logic at the
+`Stage`-level ([TODO: ref]). The main reason was that we needed to "break" the
+recursion in various domain-specific points (e.g. stages, nested stacks) which
+meant that the generic logic of "traverse the entire tree and call `synthesize`"
+did not hold. In `cdk8s`, the support for [TODO: chart dependencies] required
+that the name of the output manifest will be determined based on the topologic
+order at the app level. Here again, the generic approach failed.
+
+In leu of those failures, we decided that there is no additional value in
+actually offering a synthesis mechanism at the `constructs` level. Each
+CDK-domain implements synthesis at the "right" level. This does not mean that
+specific domains can't offer a decentralized approach (i.e. call a method called
+"synthesize" on all constructs in the tree), it just means that this is not
+provided at the base `Construct` class.
+
+### Implications on CDK code base
+
+In the AWS CDK itself, `synthesize()` was used in three locations:
+
+1. `Stack` - creates a CFN template and adds itself to the cloud assembly manifest.
+2. `AssetStaging` - stages an asset into the cloud assembly.
+3. `TreeMetadata` - creates `tree.json` with information about the construct tree.
+
+For `Stack` and `TreeMetadata`, we will convert the generic `synthesize()`
+method to `_synthesizeTemplate()` and `_synthesizeTree()` and will call them
+from the centralized synthesis function.
+
+The `AssetStaging` construct does not really need to wait until synthesis in
+order to stage the asset. In fact, all the information required already exists
+during initialization. The only missing information is the cloud assembly output
+directory, and this information is actually known during initialization (we know
+this as soon as the CDK app is created). Therefore, the solution for
+`AssetStaging` is to move the staging logic to the constructor and use
+`Stage.of(this).outdir` to find the output directory of the current stage.
```

Comment:
I wish we capitalized that `D` (`outDir`)

Reply:
We already use this capitalization in `CloudAssemblyBuilder` and I don't see this warrants a breaking change at this point.